dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -40,108 +62,210 @@ import qualified Text.Megaparsec.Pos                   as SourcePos
 -- >>> :set -XOverloadedStrings
 -- >>> import Dhall.Core (Directory (..))
 
--- | Given a Dhall expression return all imports with their location on the file.
---   Contents are already sorted by 'Src', which allows 'renderAsHtml' to traverse
---   the return of this function to inject anchors
-getImports :: Expr Src Import -> [(Src, Import)]
-getImports (Note src (Embed a)) = [(src, a)]
-getImports expr = concatMap getImports $ Lens.toListOf Core.subExpressions expr
+-- | Get the source line and column from a 'SourcePos' as an 'Int'
+getSourceLine, getSourceColumn :: SourcePos -> Int
+getSourceLine = SourcePos.unPos . SourcePos.sourceLine
+getSourceColumn = SourcePos.unPos . SourcePos.sourceColumn
+
+-- | To make each variable unique we record the source position where it was
+--   found.
+data VarDecl = VarDecl Src Text
+
+makeHtmlId :: VarDecl -> Text
+makeHtmlId (VarDecl Src{srcStart} _) =
+       "var"
+    <> Text.pack (show $ getSourceLine srcStart) <> "-"
+    <> Text.pack (show $ getSourceColumn srcStart)
+
+-- | Available ways of rendering source code as HTML
+data SourceCodeType
+    -- | Relative and remote imports are rendered using an HTML anchor tag.
+    --   Other imports are rendered as plain-text
+    = ImportExpr Import
+
+    -- | Used to render a variabled declared in let-binding that is used in
+    --   any expression
+    | VariableUse VarDecl
+
+    -- | Used to render the declaration of a variable. This is used to jump
+    --   to that variable after clicking an 'VariableUse'
+    | VariableDeclaration VarDecl
+
+{-| The 'Expr Src Import' parsed from a 'Text' is splitted into a
+    '[SourceCodeFragment]'.
+-}
+data SourceCodeFragment =
+    SourceCodeFragment
+        Src -- ^ The start and end position of this fragment
+        SourceCodeType -- ^ The type of 'SourceCodeFragment' that will guide HTML rendering
+
+-- | Returns all 'SourceCodeFragment's in lexicographic order i.e. in the same
+--   order as in the source code.
+fragments :: Expr Src Import -> [SourceCodeFragment]
+fragments = go Context.empty
+  where
+    go context = \case
+        -- The parsed text of the import is located in it's `Note` constructor
+        Note src (Embed a) -> [SourceCodeFragment src $ ImportExpr a]
+
+        -- FIXME: This doesn't work properly when the variable is quoted with
+        -- backticks since the AST records those without the quotes
```

Comment:
What precisely is the problem? Could you give an example?

Reply:
Although I know how to fix it, I can give you an example.

Suppose the following program:
```dhall
let `quoted var` = 1
in `quoted var
```
The AST will be something like this:

```
Let (Binding ... "quoted var" ...) (Note "`quoted var`" (Var "quoted var' 0))
```

For variable declarations, we don't have the exact parsed text, which leads to the following text rendering on the HTML:

```dhall
let quoted varr` = 1
in `quoted var`
```

since the range of the text computed on `fragments` doesn't take into accounts yet variable with backticks.

My idea to solve this can calculate it from using `bindingSrc0` and `bindingSrc1` `SourcePos`es. I'm going to try it and update here accordingly

Reply:
It worked like a charm and I added another test showcasing that